org.eclipse.scanning.sequencer/src/org/eclipse/scanning/sequencer/nexus/SolsticeScanMonitor.java

```diff
@@ -0,0 +1,236 @@
+		// Note: we don't use IRunListener.runPerformed as other run listeners expect the
+		// file to be closed when that method is called
+		final Dataset dataset = DatasetFactory.createFromObject(IntegerDataset.class, 1, null);
+		try {
+			this.scanFinished.setSlice(null, dataset,
+					new int[] { 0 }, new int[] { 1 }, new int[] { 1 });
+		} catch (Exception e) {
+			throw new ScanningException("Could not write scanFinished to NeXus file", e);
+		}
+	}
+
+	@Override
+	public Object getPosition() throws Exception {
+		return null;
+	}
+
+	@Override
+	public void setPosition(Object value, IPosition position) throws Exception {
+		writePosition(position);
+	}
+
+	/**
+	 * For each device, if that device writes to an external file, create an external link
+	 * within the unique keys collection to the unique keys dataset in that file.
+	 * The unique keys are required in order for live processing to take place - we need to know
+	 * how much data has been written to each file - devices may flush their data to file at
+	 * different times.
+	 * @param uniqueKeysCollection unique keys collection to add any external links to
+	 */
+	private void addLinksToExternalFiles(final NXcollection uniqueKeysCollection) {
+		if (nexusObjectProviders == null) throw new IllegalStateException("nexusObjectProviders not set");
+		
+		for (NexusObjectProvider<?> nexusObjectProvider : nexusObjectProviders) {
+			String uniqueKeysPath = (String) nexusObjectProvider.getPropertyValue(PROPERTY_NAME_UNIQUE_KEYS_PATH);
+			if (uniqueKeysPath != null) {
+				for (String externalFileName : nexusObjectProvider.getExternalFileNames()) {
+					addLinkToExternalFile(uniqueKeysCollection, externalFileName, uniqueKeysPath);
+				}
+			}
+		}
+	}
+
+	/**
+	 * Add a link to the unique keys dataset of the external file if not already present.
+	 * @param uniqueKeysCollection unique keys collection to add link to
+	 * @param externalFileName name of external file
+	 * @param uniqueKeysPath path to unique keys dataset in external file
+	 */
+	private void addLinkToExternalFile(final NXcollection uniqueKeysCollection,
+			String externalFileName, String uniqueKeysPath) {
+		// we just use the final segment of the file name as the dataset name,
+		// This assumes that we won't have files with the same name in different dirs
+		// Note: the name doesn't matter for processing purposes 
+		String datasetName = new File(externalFileName).getName();
+		if (uniqueKeysCollection.getSymbolicNode(datasetName) == null) {
+			uniqueKeysCollection.addExternalLink(datasetName, externalFileName,
+					uniqueKeysPath);
+		}
+	}
+
+	/**
+	 * Write the given position to the NexusFile.
+	 * The unique key of the position is added to the <code>uniqueKeys</code> dataset,
+	 * and the position as a string, as returned by {@link IPosition#toString()},
+	 * is added to the <code>points</code>
+	 * @param position
+	 * @throws Exception
+	 */
+	private void writePosition(IPosition position) throws Exception {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') Define and throw a dedicated exception instead of using a generic one. [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS00112)